packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -0,0 +1,318 @@
+
+	/**
+	 * Whether or not generated autoloader considers the class map authoritative.
+	 *
+	 * @var bool Whether or not generated autoloader considers the class map authoritative.
+	 */
+	private $classMapAuthoritative = false;
+
+	/**
+	 * Instantiate an AutoloadGenerator object.
+	 *
+	 * @param IOInterface $io IO object.
+	 */
+	public function __construct( $io ) {
+		$this->io = $io;
+	}
+
+	/**
+	 * Dump the autoloader.
+	 *
+	 * @param Config                       $config Config object.
+	 * @param InstalledRepositoryInterface $localRepo Installed Reposetories object.
+	 * @param PackageInterface             $mainPackage Main Package object.
+	 * @param InstallationManager          $installationManager Manager for installing packages.
+	 * @param string                       $targetDir Path to the current target directory.
+	 * @param bool                         $scanPsr0Packages Whether to search for packages. Currently hard coded to always be false.
+	 * @param string                       $suffix The autoloader suffix ignored since we want our autolaoder to only be included once.
+	 */
+	public function dump(
+		Config $config,
+		InstalledRepositoryInterface $localRepo,
+		PackageInterface $mainPackage,
+		InstallationManager $installationManager,
+		$targetDir,
+		$scanPsr0Packages = null, // Not used we always optimize.
+		$suffix = null // Not used since we create our own autoloader.
+	) {
+
+		$filesystem = new Filesystem();
+		$filesystem->ensureDirectoryExists( $config->get( 'vendor-dir' ) );
+
+		$basePath   = $filesystem->normalizePath( realpath( getcwd() ) );
+		$vendorPath = $filesystem->normalizePath( realpath( $config->get( 'vendor-dir' ) ) );
+		$targetDir  = $vendorPath . '/' . $targetDir;
+		$filesystem->ensureDirectoryExists( $targetDir );
+
+		$vendorPathCode = $filesystem->findShortestPathCode( realpath( $targetDir ), $vendorPath, true );
+
+		$appBaseDirCode = $filesystem->findShortestPathCode( $vendorPath, $basePath, true );
+		$appBaseDirCode = str_replace( '__DIR__', '$vendorDir', $appBaseDirCode );
+
+		$packageMap = $this->buildPackageMap( $installationManager, $mainPackage, $localRepo->getCanonicalPackages() );
+		$autoloads  = $this->parseAutoloads( $packageMap, $mainPackage );
+
+		$classMap = $this->getClassMapFromAutoloads( $autoloads, $filesystem, $vendorPath, $basePath );
+
+		// Write out the autoload_classmap_package.php file.
+		$classmapFile  = <<<EOF
+<?php
+
+// autoload_classmap_package.php @generated by automattic/autoload
+
+\$vendorDir = $vendorPathCode;
+\$baseDir = $appBaseDirCode;
+
+
+EOF;
+		$classmapFile .= 'return ' . $this->classMapToPHPArrayString( $classMap );
+		file_put_contents( $targetDir . '/autoload_classmap_package.php', $classmapFile );
+
+		// Copy over the autoload.php file.
+		$sourceLoader = fopen( __DIR__ . '/autoload.php', 'r' );
+		$targetLoader = fopen( $vendorPath . '/autoload_packages.php', 'w+' );
+		fwrite( $targetLoader, stream_get_contents( $sourceLoader ) );
```

Comment:
Consider using file_get_contents and file_put_contents instead of fopen and fwrite